mava/systems/jax/config.py

```diff
@@ -0,0 +1,139 @@
+# python3
+# Copyright 2021 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Config class for Mava systems"""
+
+from dataclasses import is_dataclass
+from types import SimpleNamespace
+from typing import Any, Dict, List
+
+from mava.utils.config_utils import flatten_dict
+
+
+class Config:
+    """Config handler for Jax-based Mava systems."""
+
+    def __init__(self) -> None:
+        """Initialise config"""
+        self._config: Dict = {}
+        self._current_params: List = []
```

Comment:
Do we need `self._current_params`, if we can just get the keys from `self._config.keys()`? Are they serving a different purpose? 

If they are the same, we can do `self._current_params = self._config.keys()`  and then we don't have to update `self._current_params`.

Reply:
They serve slightly different purposes. The `config` is a dictionary of dataclasses where each dataclass contains a parameter keys and values, whereas the `current_params` maintains a dict of the unwrapped config, i.e. all leaf parameter keys and values from the config. The reason we need to update the `current_params` is for when new components are added/updated we need to check if all leaf parameters still have unique names, i.e. we haven't by accident overwritten a leaf parameter by adding a new dataclass that shares a parameter name with another dataclass. I hope this makes?

Reply:
Can't we just redo calculating the list of parameters without needing `_current_params`? It might simplify the code with no real extra computational overhead?

Reply:
I don't think this would be more efficient since it would require flattening the config to obtain the leaf parameters every time a component is added/updated. I think keeping a temp list is easiest and more efficient, but there might be better ways still. Will think a bit about it.

Reply:
Sounds good :+1: